### ISSUE TITLE
Video delay, and visualisations (and playback speed)

### ISSUE THREAD

**Issue by RJVB** [zaps166/QMPlay2#750]
I just noticed that visualisations do not take the configured delay into account, causing a noticeable lag when listening over my BT speaker. It's a nice speaker for the place it's in (Klipsch The One Mk2) but it uses the basic audio codec, so I have to set a "video delay" (it's actually a transport delay of course) of +0.3 seconds.

More importantly, I wonder if there's something that can be done to improve the experience at the beginning.
- In videos where the sound track plays at frame 0 it sounds as if a tiny bit of sound gets lost when first starting the video
- Curiously, the video always starts first

That latter symptom is something I cannot really get my head around *when the correct transport has been set*, and if I were to guess I'd say that this initial bit of "silent movie" is actually longer than those 0.3 seconds. 

EDIT: something strange goes on when reducing playback speed, regardless of whether this is done keeping the audio pitch constant or not. At 0.75x speed I need to configure a transport delay of about 0.1 seconds only, and at 0.5x it has to be set to 0.
It looks like speeding up the playback doesn't require changing the transport delay though of course it becomes increasingly difficult to detect sync issues.

**Comment by zaps166** [comment]
Bluetooth speakers has variable audio latency - depends on codec and on signal strength - low signal might increase latency because of larger buffer to prevent audio stuttering (and also can lower the bitrate). 

Use cable whenever possible (many BT speakers has JACK input). With cable you're always full quality, lowest and fixed latencies and you can see your "signal" range.

Visualizations doesn't respect video latency - I don't want to buffer it there and complicate things. It's not video anyway :smile: 

About playback speed - I don't know, but I think when speed is too low, the latency doesn't matter that much. When speed is too high, you should still hear the latency. Also rubberband filter has latency, but it should be implicitly added to video delay.

**Comment by RJVB** [comment]
>Use cable whenever possible (many BT speakers has JACK input). With cable you're always full quality, lowest and fixed latencies and you can see your "signal" range.

Well, DOH, but a cable isn't always feasible, is it. Can't just span one the middle of a (kitchen) table across the space to where my speaker sits...

>Visualizations doesn't respect video latency - I don't want to buffer it there and complicate things. It's not video anyway :smile: 

For visualisations it's one thing, for the FFT spectrum another, IMHO. I don't know how you implemented the former, but I suspect you have to buffer for the latter anyway.

>About playback speed - I don't know, but I think when speed is too low, the latency doesn't matter that much.

I don't see how that could make sense, unless the slowing down introduces its own delay w.r.t. the audio. Which might be the case since there is indeed a noticeable sync problem when I play the audio of a slowed-down video over the builtin speakers.

>When speed is too high, you should still hear the latency. Also rubberband filter has latency, but it should be implicitly added to video delay.

Is the rubberband filter always in the loop, with just an on/off switch to determine whether it does something to pitch? Because that noticeable sync problem mentioned above is there whether I keep audio pitch constant or not.

I think I had actually noticed it before, as the slowed-down playback isn't as helpful in figuring out what musicians are doing with their hands (which is what I hoped to use the constant-pitch feature for).

I'm attaching the audio/video sync test I use - and just FWIW: visualisations and FFT appear to be completely out of sync with the audio for me.

https://github.com/user-attachments/assets/d106101a-c14f-41c5-b1bb-407fe583d8de

**Comment by zaps166** [comment]
For me this video plays perfect in QMPlay2 regardless of keep pitch and playback speed (using PC's sound card and pipewire).

> Is the rubberband filter always in the loop, with just an on/off switch to determine whether it does something to pitch? 

Rubberband filter is running only when your playback speed != 1.0 and you have keep pich enabled. Otherwise it's unused, resources are freed.

> visualisations and FFT appear to be completely out of sync with the audio for me.

How about video? As I told before - video delay parameter doesn't work in visualizations .

> For visualisations it's one thing, for the FFT spectrum another, IMHO.

What do you mean? The only buffer there is to grab enough samples to display and/or calculate FFT. They should display at same time (I mean some milliseconds doesn't matter like 1-2x screen refresh rate or so).

**Comment by RJVB** [comment]
>For me this video plays perfect in QMPlay2 regardless of keep pitch and playback speed (using PC's sound card and pipewire).

I suppose your PC is a lot faster than mine, but is there a reason to expect that slowing down the video increases computational load such that video playback becomes delayed w.r.t. audio playback? On the Linux system where I observed the sync issues I actually have to dial in a negative video delay when I slow down playback to 0.25x! 
I'll see what happens on my Mac.

>> For visualisations it's one thing, for the FFT spectrum another, IMHO.
>
>What do you mean?

I consider visualisations mostly eye candy, but the FFT spectrum is supposed to correspond to what you're hearing. Showing it for something else doesn't really make sense (not much more than showing a random spectrum).

>The only buffer there is to grab enough samples to display and/or calculate FFT.

Let's just stick to the situation where the video delay setting is used to dial in a transport delay in the audio playback (so it's a positive delay). In that case, can't you simply fill the buffer for calculating the FFT with audio samples from X seconds ago? Don't you keep any decoded audio data after it has been handed off to the audio device?

>They should display at same time (I mean some milliseconds doesn't matter like 1-2x screen refresh rate or so).

Who are "they" here?

( 1 screen refresh is already almost 17 ms for a 60Hz display ;) )

**Comment by RJVB** [comment]
On Friday July 12 2024 13:12:52 Błażej Szczygieł wrote:
>For me this video plays perfect in QMPlay2 regardless of keep pitch and playback speed (using PC's sound card and pipewire).

Basically idem on my Mac. At some point I thought I noticed a slight offset at 0.25x speed but couldn't reproduce that.
A few important differences of course:
- VTB acceleration rather than VAAPI
- OpenGL rendering instead of Vulkan
- "direct" audio rendering rather than via Pulseaudio.

I did manage to build Pipewire on my Linux rig but haven't yet figured out how to start it instead of Pulseaudio. (I tried once to launch it in addition to assuming the daemons shouldn't bite but it messed up the sound volume levels until I restarted the PA daemon.)

Did you notice that QMPlay2 identifies the Twitch video as 120Hz in the Info panel instead of the actual 60Hz? `ffprobe` shows it's also a "120 tbr" (??) video track; maybe that's where the confusion comes from?

**Comment by zaps166** [comment]
Notes about Bluetooth audio: yes, this method has variable audio latency (depends on devices, codec used, signal strength). You can compensate it manually for video playback in QMPlay2, but not for visualizations. Visualizations in QMPlay2 doesn't have additional buffering, so it's impossible. Adding this is possible, but too much work and too much code for this use-case. Sorry, I can't complete every feature request :disappointed: 

Also bluetooth audio latency can change at "random" in some cases, so you'll have to compensate it again manually.

---

Once I had two Bluetooth mono speakers (no JACK cable input...) and I wanted to use them as stereo. After manual latency synchronization between them (in PipeWire), they played quite well. After some time they desynced and echo appeared again.

**Comment by RJVB** [comment]
If memory serves me right the main thing here is the fact that I often lose some sound at the beginning of a video started with latency compensation, esp. noticeable in music videos where the music starts at the 1st frame. It's as if the audio is being delayed, which is the opposite of what should happen.

Or is this related to bluetooth too? IOW, would it help to "prime" the audio channel maybe 1s before playback starts - which could be hidden easily in the network delays when starting a YT video?

**Comment by zaps166** [comment]
Hmm. I experience it on HDMI when HDMI is in... power-saving mode? One it wakes-up, it plays normally, but usually I lost begin of audio. Maybe same with bluetooth? Is your issue related to QMPlay2 or it's a general issue?

**Comment by RJVB** [comment]
I have the impression it's more noticeable when playing YT videos using QMPlay2's latency correction. In this video e.g. (https://www.youtube.com/watch?v=soDuDgnqjE4) the 1st note swells into existence after a short initial silence. When I just start the playback with +0.3s latency it starts abruptly, also when I stop and restart the video so nothing is supposedly gone into standby. But when I just drag the cursor back to the beginning I do hear that initial silence and the swelling of the note.

I'm still using Pulseaudio and don't even know if the latency you can configure there (via e.g. pavucontrol) actually does something - it certainly doesn't interfere with QMplay2's correction (QP2 playback uses an "ALSA Plugin"). I'm also using a USB BT DAC (Sennheiser BTD600) which may add its own side-effects?)

**Comment by zaps166** [comment]
> QP2 playback uses an "ALSA Plugin"

I think for PulseAudio you should use PulseAudio plugin. ALSA is for `hw:` or when use use their `dmix` (no need to use ALSA API to pass audio to PulseAudio - I think better directly).

> I'm still using Pulseaudio

I forgot about it long time ago and my sound issues are gone with PipeWire.

> the 1st note swells into existence after a short initial silence. When I just start the playback with +0.3s latency it starts abruptly, also when I stop and restart the video so nothing is supposedly gone into standby. But when I just drag the cursor back to the beginning I do hear that initial silence and the swelling of the note.

Exactly same here on HDMI regardless of player used. I noticed that the latency when audio starts to be hear depends on connected hardware (sometimes is 1 second, sometimes more). I rarely use HDMI, so workaround is unknown so far. If you add additional latency via pavucontrol, you simply has more time to "wait" for the audio device to be ready to output sound. I still don't know if it's hardware issue or driver issue.

**Comment by zaps166** [comment]
> it certainly doesn't interfere with QMplay2's correction

QMPlay2 corrects A/V sync only.

---

You need to add about 0.5 seconds silence on playback start (when your audio device is open or resumed) to workaround hardware/driver issue.

**Comment by RJVB** [comment]
On Saturday December 14 2024 15:47:05 Błażej Szczygieł wrote:
>> it certainly doesn't interfere with QMplay2's correction

>QMPlay2 corrects A/V sync only.

Hmmm, of course, I was thinking of additional compensation added by downstream layers. But a transport delay in the audio channel can only be compensated by delaying the video and I don't see how PA could do that in an application that handles video playback itself.
I keep the latency correction configured via pavucontrol because it seems this improves sync in in-browser content playback - possibly because the browser queries the value from PA and adds the corresponding delay to the appropriate channel?

I have a vague memory that I kept QMPlay2 set to ALSA/PA Sound Server because the PA ALSA plugin ignores the latency setting, but the above explanation seems more plausible.

Still, after some more testing right now it does seem like the PA latency offset is to blame?!

>You need to add about 0.5 seconds silence on playback start (when your audio device is open or resumed) to workaround hardware/driver issue.

Play 0.5s of silence (more or less what I called priming) before actually starting the content playback? That's currently not possible to achieve, is it?
And that still doesn't explain why the phenomenon is more noticeable when correcting for an audio transport delay! It seems to me that you shouldn't even notice the correction when the latency is set correctly (and when you switch to the internal audio device with 0 latency you should see a short delay in the video). Again, right now it's more like the video starts playing without sound, and then all of a sudden sound starts.

Same with a negative or a positive PA latency offset, and that just doesn't seem to make sense.

I do have pipewire installed nowadays, including its pulseaudio drop-in replacement. I haven't yet figured out how to start it correctly; the one time I did try it messed up the max. volume I could get through PA when I went back to that.

**Comment by zaps166** [comment]
> Still, after some more testing right now it does seem like the PA latency offset is to blame?!

Latency in pavucontrol shouldn't be used for your issue. It may generate other issues.

---

Found solution, it's [here](https://wiki.archlinux.org/title/PipeWire#Noticeable_audio_delay_or_audible_pop/crack_when_starting_playback) for PipeWire with WirePlumber >= 0.5.0.

You shouldn't set it for all devices, just set for your device (you can use `pw-top` to obtain the name) or maybe set the config for bluetooth devices only. `pw-top` will show you is it suspends or not.

My config for HDMI (`~/.config/wireplumber/wireplumber.conf.d/51-disable-suspension.conf`):
```
monitor.alsa.rules = [
  {
    matches = [
      {
        node.name = "~alsa_output.*.hdmi*"
      }
    ]
    actions = {
      update-props = {
        session.suspend-timeout-seconds = 0
      }
    }
  }
]
```

Your first audio playback in system will be cut, but every other will not cut the sound.

**Comment by RJVB** [comment]
On Sunday December 15 2024 00:57:36 Błażej Szczygieł wrote:

>> Still, after some more testing right now it does seem like the PA latency offset is to blame?!
>Latency in pavucontrol shouldn't be used for your issue. It may generate other issues.

I agree, at least not for compensating an audio transport delay by adding yet more delay ;) But it clearly does do something. Guess I'll have to RTFM ;)

>Found solution, it's [here](https://wiki.archlinux.org/title/PipeWire#Noticeable_audio_delay_or_audible_pop/crack_when_starting_playback) for PipeWire with WirePlumber >= 0.5.0.
>
>You shouldn't set it for all devices, just set for your device (you can use `pw-top` to obtain the name) or maybe set the config for bluetooth devices only. `pw-top` will show you is it suspends or not.

Worth a try if I can figure out how to switch to pipewire without breaking anything, including my PulseAudio config.

Meanwhile, my BT device is seen as a USB DAC by the system, and the LEDs on it go a bit duller after about 5 seconds of not receiving any signal.

Also, I can reproduce the issue with the internal sound device too. Not for small latency offsets nor for long ones >= 900ms or so, so it's not something like a glitch in a different code path that's taken when lat.offset!=0 .

>Your first audio playback in system will be cut, but every other will not cut the sound.

Sorry, I can't make sense of that. You mean the 1st time I play audio, or the 1st device in the list?

**Comment by zaps166** [comment]
> Meanwhile, my BT device is seen as a USB DAC by the system, and the LEDs on it go a bit duller after about 5 seconds of not receiving any signal.

Yes, this is probably the reason why you can't hear begin of the audio - the device is waking-up from sleep. If it's a USB DAC, just set `node.name` according to what `pw-top` sees on your device.

> Also, I can reproduce the issue with the internal sound device too. Not for small latency offsets nor for long ones >= 900ms or so, so it's not something like a glitch in a different code path that's taken when lat.offset!=0 .

:thinking: Don't understand.

> Sorry, I can't make sense of that. You mean the 1st time I play audio, or the 1st device in the list?

I mean - when you start any playback on your device when suspend is disabled in PipeWire - the device will not go sleep anymore, so it'll work. However the device has to be accessed at least once to wake-up, so first audio you play after device connection or PC boot - begin of the audio will not be emited (only once).

---

Also you can try (not working for HDMI to me):
```sh
echo 0 | sudo tee /sys/module/snd_hda_intel/parameters/power_save
echo 0 | sudo tee /sys/module/snd_hda_intel/parameters/power_save_controller
```

**Comment by RJVB** [comment]
On Sunday December 15 2024 07:13:32 Błażej Szczygieł wrote:

>Yes, this is probably the reason why you can't hear begin of the audio - the device is waking-up from sleep. If it's a USB DAC, just set `node.name` according to what `pw-top` sees on your device.

I'm not convinced, because I can see the device wake up while QMPlay2 still seems to be starting the playback. There may be the tiniest bit of an artefact when I start playback from the standby state without latency correction but it's definitely more pronounced when there is a latency correction.

Would it be an invasive change to open the configured audio channel (and send "something" into it?) as soon as the user hits play, if that's not already being done? That might fix or alleviate the issue in >75% of the cases. And I'm guessing the same issue will exist on platforms where you can't configure suspension like you showed for pipewire

Except on Mac with bitperfect playback ... that clearly wakes the audio device up :)

>> Also, I can reproduce the issue with the internal sound device too. Not for small latency offsets nor for long ones >= 900ms or so, so it's not something like a glitch in a different code path that's taken when lat.offset!=0 .
>
>:thinking: Don't understand.

You don't understand my words or what I describe?

>However the device has to be accessed at least once to wake-up, so first audio you play after device connection or PC boot - begin of the audio will not be emited (only once).

Yes, that makes sense.

>Also you can try (not working for HDMI to me):
>```sh
>echo 0 | sudo tee /sys/module/snd_hda_intel/parameters/power_save
>echo 0 | sudo tee /sys/module/snd_hda_intel/parameters/power_save_controller
>```

I must already have set those ("N" for the controller but that's surely the same thing), if it's not the default. Possibly when trying to minimise the delay in notification sounds.

**Comment by zaps166** [comment]
> You don't understand my words or what I describe?

On your internal sound - are you adding latency offset in pavucontrol? Do you have any issues when you don't add any latency in pavucontrol?

> I'm not convinced, because I can see the device wake up while QMPlay2 still seems to be starting the playback. There may be the tiniest bit of an artefact when I start playback from the standby state without latency correction but it's definitely more pronounced when there is a latency correction.

I don't know, I can't see/touch this :smile: Maybe it's waiting and after this still not playing from the begin. I have only HDMI, same behavior (all applications, not only QMPlay2). Sometimes it "cuts" 1 sec of audio, sometimes 4 sec of audio...

> Would it be an invasive change to open the configured audio channel (and send "something" into it?) as soon as the user hits play, if that's not already being done? That might fix or alleviate the issue in >75% of the cases. And I'm guessing the same issue will exist on platforms where you can't configure suspension like you showed for pipewire

I don't know how long the silence must be sent - it depends on hardware/driver. Also it must be send on every resume from pause, because I don't know when device really suspends (maybe can ask PipeWire for suspend time, but it's not universal way and much more complicated). And finally it'll be done for QMPlay2 only, other software will still have this problem.

If you just push 1 sec silence before first audio write, QMPlay2 will discard begin on audio due to video sync.

**Comment by RJVB** [comment]
On Sunday December 15 2024 10:32:55 Błażej Szczygieł wrote:
>> You don't understand my words or what I describe?
>
>On your internal sound - are you adding latency offset in pavucontrol? 

I did only for testing if I could reproduce the issue that way.

>Do you have any issues when you don't add any latency in pavucontrol?

Nope. And based on what you've been suggesting that's probably because the internal DAC is configured not to resume.

>Also it must be send on every resume from pause, because I don't know when device really suspends 

Ah, yes! I guess I must be expecting glitches when resuming from pause because the effect doesn't bother me there. There's of course also the effect that it's not trivial to reproduce the exact glitch by pausing at the exact same time.

But isn't there a way to keep the device from suspending when playback is paused?

>And finally it'll be done for QMPlay2 only, other software will still have this problem.

Well, that's fine with me - I'm using QMPlay2 for low-resource, high-audio-quality playback (except on MSWin when I use Foobar2000 for audio).

>If you just push 1 sec silence before first audio write, QMPlay2 will discard begin on audio due to video sync.

Even if you pretend the video also starts with the same duration of something else - say a dissolve effect of the rest screen?

**Comment by zaps166** [comment]
> Even if you pretend the video also starts with the same duration of something else - say a dissolve effect of the rest screen?

Video needs to be informed about it somehow. Need to look at code, don't remember (also it's very old code).

> except on MSWin when I use Foobar2000 for audio

:smile: 

> But isn't there a way to keep the device from suspending when playback is paused?

PipeWire metod as I gave you before (also PulseAudio can do this).

QMPlay2 - need to play silence all the time when paused.

---

Alternative - play silence somewhere in background process all the time - it'll never suspend.

**Comment by RJVB** [comment]
>PipeWire metod as I gave you before (also PulseAudio can do this).

Nothing to do with delays or latencies, but I just read that PW resamples everything to a "central frequency" and from there to whatever the output device is set to.

Unless PA does the same thing that doesn't sound like a recipe for getting better output quality nor shorter latencies...?!

In short, neither system will do anything to get even close to bit-perfect transfer? I thought PA would switch to its alternate frequency (usually to 48k from 44.1k) if that allowed an integer factor resampling but it looks like it isn't doing that on my system.

**Comment by zaps166** [comment]
> Nothing to do with delays or latencies, but I just read that PW resamples everything to a "central frequency" and from there to whatever the output device is set to.

You can bypass PipeWire and use ALSA output in QMPlay2 set to `hw:n` to use the device directly, so you'll have bit-perfect exclusive access (all other streams must shutdown before playback). "Force samplerate" in QMPlay2 **must** be disabled.

Another solution is to create config in `/etc/pipewire/pipewire.conf.d/` and set array of allowed frequencies (set freq you want, you can set more than one):
```
context.properties = {
     default.clock.allowed-rates = [ 48000 ]
}
```

Restart services and observe `pw-top` what will happen.

**Comment by RJVB** [comment]
On Thursday December 19 2024 12:44:56 Błażej Szczygieł wrote:

>You can bypass PipeWire and use ALSA output in QMPlay2 set to `hw:n` to use the device directly

Does that work with PW? With PA you can't unless you exclude the device you want. I think it has to do with the compulsory "monitor of XXX" devices.

>Another solution is to create config in `/etc/pipewire/pipewire.conf.d/` and set array of allowed frequencies (set freq you want, you can set more than one):
>```
>context.properties = {
>     default.clock.allowed-rates = [ 48000 ]
>}

I read about that, but 1) there's apparently a kernel bug that hasn't been fixed in the kernel I'm using. IIUC it has to do with recording at a different rate than playing back, which I'd never do so it may not be relevant. And 2) I haven't yet seen any indication that this would bypass to initial resampling to some internal rate.

**Comment by zaps166** [comment]
Try this command `watch -t -n 1 cat /proc/asound/card*/pcm*p/sub*/hw_params` and see sample rates - it comes from audio driver.

Also open `pw-top` - it shows what PipeWire is doing with samplerate. If all are the same - I guess there's no resampling from the player to the audio device.

---

> Does that work with PW? With PA you can't unless you exclude the device you want. I think it has to do with the compulsory "monitor of XXX" devices.

If PW closes audio device (suspend must be enabled), you can open it using ALSA plugin using `hw:n,m` device in settings. It's direct access to the device, it skips PipeWire/JACK/Pulseaudio/whatever completely.

**Comment by RJVB** [comment]
On Friday December 20 2024 09:38:47 Błażej Szczygieł wrote:
>If all are the same - I guess there's no resampling from the player to the audio device.

That's what I'd hope too, but I suppose it depends what that internal frequency is - probably not 2116mHz (44.1k * 48k) ;)

>If PW closes audio device (suspend must be enabled), you can open it using ALSA plugin using `hw:n,m` device in settings. It's direct access to the device, it skips PipeWire/JACK/Pulseaudio/whatever completely.

I saw that, yes. Should work like that in PA too, but I guess it's those pesky monitors that wreak havoc.

**Comment by zaps166** [comment]
> That's what I'd hope too, but I suppose it depends what that internal frequency is - probably not 2116mHz (44.1k * 48k) ;)

What driver or hardware is doing is - let's say - out of scope. I'm talking about not using PipeWire/PulseAudio for this. Or configure PipeWire/PulseAudio to use different sample rate to match your audio file (so it'll depend on driver/hardware, PipeWire will not resample).

---

This ALSA settings works for my integrated audio device:

![Zrzut ekranu_20241221_092107](https://github.com/user-attachments/assets/8c193f00-f416-4a2c-b0d8-54921bd8e65e)

In this case, PipeWire is not used:

![Zrzut ekranu_20241221_092444](https://github.com/user-attachments/assets/00676fb1-5bc5-4bdd-afd4-02045ff73955)

192kHz stereo samplerate in QMPlay2, the driver `hw_params`:

![Zrzut ekranu_20241221_092215](https://github.com/user-attachments/assets/6e13107f-8ce8-486b-9c78-552d58a2d93c)

I can measure high frequencies (generated by Tone generator in QMPlay2) e.g. 80kHz on JACK output, so it's really 192kHz (or around 192kHz - idk what driver/hardware is doing).

**Comment by RJVB** [comment]
reviving this for now rather than filing a new ticket:

Now that I can use direct ALSA access to my BT speaker I notice that even with this method I still get this phenomenon where the video starts to play and audio kicks in some time later when I configure the `T=0.3s` delay necessary to get get acceptable audio/video sync.

In the sync-testing video above this boils down to getting the test sound only at the 2nd time the cursor passes the centre (0 point). Even when the DAC has been primed, e.g. when I pause the video, drag the playback cursor back to the beginning and unpause.

I still think that this suggests that audio playback is not started `T` seconds ahead of the video playback, but after our recent discussion about how you determine actual frame rates I wonder if that doesn't have something to do with what I'm observing here?

Direct ALSA access means there's no sound daemon process  that can explain delays, only delays in QMPlay2 itself (multithreading?) because it isn't registered with rtkit like PulseAudio is?

**Comment by zaps166** [comment]
QMPlay2 doesn't read audio sink real latency (it's hard to smoothly adjust video to audio when latency varies from frame to frame), but it applies latency compensation depends on fixed latency setting which is more or less true for sound cards.

Bluetooth has higher latencies and **variable** latencies, so it's expected. There's no latencies issues on sound cards, so it's working correct in normal scenarios.

> Direct ALSA

Bluetooth audio goes through BlueZ, so ALSA is only pass-through for BlueZ.

> only delays in QMPlay2 itself (multithreading?) because it isn't registered with rtkit like PulseAudio is?

Latency depends on buffer size set in audio sink API, whether equalizer or other filters are used, decoded audio chunks (audio codec latency) - but these should be mostly compensated when playing video. What is not compensated with video sync in QMPlay2 is latency between audio driver and hardware (which in case of Bluetooth is high and variable - also depends on signal quality and codec used there).

Threading and priorities doesn't matter - we're not talking about 1ms, but about 300ms and it doesn't matter - when video is synced with this latency, it still plays correctly (it's not game or real time audio processing when somebody talks to microphone for example). I think RT priorities are used to prevent audio cracking on low latencies.

---

Another issue is when DAC wakes up from suspend (common HDMI issue), so begin of audio is not emitted to speakers - in this case just prevent power saving in your system settings (it increases idle power consumption, see example: https://github.com/zaps166/QMPlay2/issues/750#issuecomment-2543574302).

**Comment by zaps166** [comment]
Also latency may increase when you have slow CPU (or GPU in case of GPU video decoding) - video is keep in sync with audio, but in edge cases the latency is increased.

**Comment by zaps166** [comment]
Notice this video has 120 FPS - check if your hardware can handle it smoothly. For test go real fullscreen (without compositor overhead) and disable V-Sync or use mailbox mode for Vulkan.

**Comment by RJVB** [comment]
On Sunday July 20 2025 10:44:59 Błażej Szczygieł wrote:

>QMPlay2 doesn't read audio sink real latency 

I'm talking about latency set via the in-app feature ("video delay"). 


>> Direct ALSA
>
>Bluetooth audio goes through BlueZ, so ALSA is only pass-through for BlueZ.

Not in this case; I'm using a BT DAC, which is seen as a standard USB DAC by ALSA and I thus do not have to worry about the Linux BT stack at all.
Still, I find that 0.3s is good enoug estimate for the average BT delay with its standard audio codec that I don't notice sync issues

>Another issue is when DAC wakes up from suspend (common HDMI issue)

That's also moot here when I pause/rewind/unpause. 

I should film the phenomenon so you can see/hear what I'm talking about.

It is of course possible that something is going on inside that DAC of mine, in fact I'm beginning to believe that might be what I'm looking at here.

I guess that QMPlay2 doesn't touch (discard) the audio buffers (or whatever) when you pause playback, but does it do that when you displace the playback cursor during the pause?

**Comment by RJVB** [comment]
>Notice this video has 120 FPS 

Oh? I'm of course testing with a local copy which claims to be 60Hz.

Either way, I don't see how this could explain why there is a considerable (on the scale of things) *advance* of the video...

**Comment by zaps166** [comment]
> Notice this video has 120 FPS
> Oh? I'm of course testing with a local copy which claims to be 60Hz.
> 
> Either way, I don't see how this could explain why there is a considerable (on the scale of things) *advance* of the video...

Right, 60 FPS, but shows 120 in file info xD

**Comment by zaps166** [comment]
> I'm talking about latency set via the in-app feature ("video delay").

Yes, you compensate it manually by adding an offset.

> Not in this case; I'm using a BT DAC, which is seen as a standard USB DAC by ALSA and I thus do not have to worry about the Linux BT stack at all.
Still, I find that 0.3s is good enoug estimate for the average BT delay with its standard audio codec that I don't notice sync issues

Right, forgot about it, but idk how it works, best to have it on my PC here to check what's really going on. Remember it's still wireless. How about `ffplay` from command line using ALSA HW access?

> That's also moot here when I pause/rewind/unpause.

With PipeWire you can fix it (PipeWire will constatly keep device open and maybe send silence there), with direct ALSA access if driver/device doesn't have such option - you can't fix it probably.

> I should film the phenomenon so you can see/hear what I'm talking about.
It is of course possible that something is going on inside that DAC of mine, in fact I'm beginning to believe that might be what I'm looking at here.
I guess that QMPlay2 doesn't touch (discard) the audio buffers (or whatever) when you pause playback, but does it do that when you displace the playback cursor during the pause?

Well, I don't see and don't hear it.

When you pause the `snd_pcm_pause()` is called if it's supported `snd_pcm_hw_params_can_pause()`, so DAC can suspend. You can look at code, put more logs and see if it's called and what'll happen when you don't call pause. Also you can try one of these functions: `snd_pcm_drain()`, `snd_pcm_drop()`.

**Comment by RJVB** [comment]
On Monday July 21 2025 11:29:05 Błażej Szczygieł wrote:

>Right, forgot about it, but idk how it works

? I am talking about the QMPlay2 feature to delay the video, not the latency setting in PulseAudio!


>Well, I don't see and don't hear it.

I tested on MSWin with WASAPI access to the BT audio stack. Here I'm getting the audio at the right time when I configure the appropriate delay but I'd swear that the video isn't being delayed but being slowed down. And only until the first sound is made. Cute but rather unexpected :)

This was with a different BT receiver and annoyingly I couldn't get the Sennheiser DAC to connect to it...

>When you pause the `snd_pcm_pause()` is called if it's supported `snd_pcm_hw_params_can_pause()`, so DAC can suspend. You can look at code, put more logs and see if it's called and what'll happen when you don't call pause. Also you can try one of these functions: `snd_pcm_drain()`, `snd_pcm_drop()`.

OK, I'll fool around with that.

**Comment by zaps166** [comment]
> ? I am talking about the QMPlay2 feature to delay the video, not the latency setting in PulseAudio!

Ikr. I replied to this few lines above. Idk how your USB adapter works.

---

How about ffplay for example?

**Comment by RJVB** [comment]
On Monday July 21 2025 11:29:05 Błażej Szczygieł wrote:
>what'll happen when you don't call pause.

Doesn't make any difference, except that there is a bigger chance that I still hear a sound after playback was stopped.

Bigger chance because if you stop playback of this particular video just before the cursor reaches the 0 position you still get to hear the test sound. Exactly as you'd expect.

BTW, yes, canPause is true for me.

>Also you can try one of these functions: `snd_pcm_drain()`, `snd_pcm_drop()`.

The logical one would be `snd_pcm_drop` to minimise the chances that audio continues to sound after pausing. But I haven't been able to figure out how to recover from calling either of the functions, so playback fails as soon as I unpause.

The funny thing is that the LEDs on the BTD600 go dim when I terminate playback, but not when I pause. This suggests that `snd_pcm_pause` doesn't actually allow the device to go into standby mode ... as you'd hope for paused playback!

**Comment by RJVB** [comment]
https://github.com/user-attachments/assets/9cbf790b-916b-4d20-989a-826ee9cec700

**Comment by zaps166** [comment]
Slow motion at start is because QMPlay2 tries to sync with 0.3 s offset. It's not perfect algorithm, but it affects start of playback only. Once you play a video, it'll play correctly after about a second.

Maximum latency in ALSA module settings should not impact video, but it'll impact visualizations, volume control, equalizer, etc. This value for ALSA HW should be as small as possible just not have crackling audio (or maybe not to have too high CPU load in some cases).

Muted audio at start sounds like DAC resume-from-suspend issue. Data which has not been played, because DAC was not yet ready, is discarded. Same on HDMI/DP devices.

Delayed and later too fast audio at resume from pause sounds like DAC resume-from-suspend issue where it still has old buffer and audio driver or device firmware tries to drop it somehow and "catch-up" the audio stream. On pause ALSA has pause command, the device is not entirely closed. Maybe driver issue or USB adapter issue (I read your message that nothing fully helped when you modified the code)?

**Comment by RJVB** [comment]
>Once you play a video, it'll play correctly after about a second.

Yes, 1s seems about right.

> This value for ALSA HW should be as small as possible just not have crackling audio (or maybe not to have too high CPU load in some cases).

Do you mean it does NOT limit the "channel delay" that can be specified via "delay video"? IOW, it does not have to be at least 0.3s if that's the delay I need for my BT DAC?

> (I read your message that nothing fully helped when you modified the code)?

Indeed, my attempts to improve things by modifications of the code didn't help at all.

I think we're going to have to write this down to a firmware issue or implementation detail which manifests itself as a startup channel delay.

The only thing I can still try to do is to check playback in my web browser. AFAICT this does use the PA latency setting that can be specific e.g. via pavucontrol. I think that has to be the browser that reads the parameter and applies it to delay video playback because I don't see how the PA daemon could do anything meaningful to compensate an audio channel delay (time warping? ;) ).

I wonder how far it would help to start sending silence to the audio device 0.3s before the video starts to play, but that's a mod that requires a little more knowledge of your code...

**Comment by zaps166** [comment]
Could you change this https://github.com/zaps166/QMPlay2/blob/ee9263a6304eae2be00649fd80da5a7069660889/src/modules/ALSA/ALSAWriter.cpp#L244 to `modParam("delay", delay_us / 1000000.0 + 0.3);` and check again?

**Comment by RJVB** [comment]
That looks (and behaves) like hard-coding the video delay that I specify manually otherwise ... correct?

**Comment by zaps166** [comment]
> That looks (and behaves) like hard-coding the video delay that I specify manually otherwise ... correct?

Yes, but is it working better than option available in UI?

**Comment by RJVB** [comment]
In fact, instead of suggesting to send 0.3s of silence before the video starts I probably should have said to start the audio playback 0.3s before the video starts. If you're doing that already then it might indeed be interesting to try "priming" the audio by sending silence.

But in that case I should probably see what Sennheiser have to say about this...

**Comment by RJVB** [comment]
>Yes, but is it working better than option available in UI?

No, exactly the same. If anything there are more artifacts when I rewind the video (though that's hard to quantify).

**Comment by zaps166** [comment]
> Yes, but is it working better than option available in UI?
> No, exactly the same. If anything there are more artifacts when I rewind the video (though that's hard to quantify).

Ok, so still slow motion at start, right?

**Comment by RJVB** [comment]
>Ok, so still slow motion at start, right?

Oh, no, probably not. But then again I hardly noticed that before on the machine I'm testing that on; it was much more evident on the (*much*) faster MSWin machine. But there I cannot get the Sennheiser DAC to work ...

**Comment by RJVB** [comment]
Here's an interesting tidbit : I thought I could emulate that idea of priming the DAC (with silence) by pausing the playback well before the 1st actual sound is made.

Tried 2 different ways: 1) by hitting <space> immediately after launching playback (thus pausing before there's even an image) and 2) by hitting space as soon as an image appears (so there's about 100ms of playback).

Neither makes a difference, and it also doesn't make a difference if I modify the code so the device isn't actually paused.

Which makes me wonder if the DAC would have to be primed with an actual sound rather than silence ...

**Comment by RJVB** [comment]
>Which makes me wonder if the DAC would have to be primed with an actual sound rather than silence ...

So I toyed a bit with this idea, using a video that has applause (basically white noise) starting at video frame 0.

When I don't do anything (except configure the required video/transport delay), I hear the audio starting at approx. 1.8 seconds.

When I pause the video immediately after starting it, that "gap" is shortened to a bit under 1.2 seconds. Hardly noticeable, but consistent when measured with a stopwatch.

I do *not* measure any changes with the test video we've been using until now.
